ocpmodels/datasets/ase_datasets.py

```diff
@@ -59,24 +66,24 @@ def __init__(self, config, transform=None, atoms_transform=apply_one_tags):
         if self.config.get("keep_in_memory", False):
             self.data_objects = {}
 
-        # Derived classes should extend this functionality to also create self.id,
+        # Derived classes should extend this functionality to also create self.ids,
         # a list of identifiers that can be passed to get_atoms_object()
 
     def __len__(self):
-        return len(self.id)
+        return len(self.ids)
 
     def __getitem__(self, idx):
         # Handle slicing
         if isinstance(idx, slice):
-            return [self[i] for i in range(len(self.id))[idx]]
+            return [self[i] for i in range(*idx.indices(len(self.ids)))]
 
         # Check if data object is already in memory
         if self.config.get("keep_in_memory", False):
```

Comment:
This functionality can be simplified significantly by using [`functools.cache`](https://docs.python.org/3/library/functools.html#functools.cache)

Reply:
Good idea! I added an optional functools.__cache__ to the __getitem__ method.

This would be a slight performance regression for the AseReadMultiStructureDataset since every structure will get read twice (once on startup to find the number of structures, and a second time when an item is loaded and cached, but this feels relatively minor.